lucene/core/src/java/org/apache/lucene/store/MemorySegmentIndexInput.java

```diff
@@ -0,0 +1,525 @@
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.lang.invoke.VarHandle;
+import java.nio.ByteOrder;
+import java.util.Arrays;
+import java.util.Objects;
+
+import org.apache.lucene.util.IOUtils;
+
+import jdk.incubator.foreign.MemoryHandles;
+import jdk.incubator.foreign.MemorySegment;
+
+/**
+ * Base IndexInput implementation that uses an array of MemorySegments to represent a file.
+ *
+ * <p>For efficiency, this class requires that the segment size are a power-of-two
+ * (<code>chunkSizePower</code>).
+ */
+public abstract class MemorySegmentIndexInput extends IndexInput implements RandomAccessInput {
+  // We pass 1L as alignment, because currently Lucene file formats are heavy unaligned: :(
+  static final VarHandle VH_getByte = MemoryHandles.varHandle(byte.class, 1L, ByteOrder.BIG_ENDIAN).withInvokeExactBehavior();
+  static final VarHandle VH_getShort = MemoryHandles.varHandle(short.class, 1L, ByteOrder.BIG_ENDIAN).withInvokeExactBehavior();
+  static final VarHandle VH_getInt = MemoryHandles.varHandle(int.class, 1L, ByteOrder.BIG_ENDIAN).withInvokeExactBehavior();
+  static final VarHandle VH_getLong = MemoryHandles.varHandle(long.class, 1L, ByteOrder.BIG_ENDIAN).withInvokeExactBehavior();
+  
+  final boolean isClone;
+  final long length;
+  final long chunkSizeMask;
+  final int chunkSizePower;
+  final MemorySegment[] segments;
+  
+  int curSegmentIndex = -1;
+  MemorySegment curSegment; // redundant for speed: segments[curSegmentIndex], also marker if closed!
+  long curPosition; // relative to curSegment, not globally
+
+  public static MemorySegmentIndexInput newInstance(
+      String resourceDescription,
+      MemorySegment[] segments,
+      long length,
+      int chunkSizePower) {
+    if (segments.length == 1) {
+      return new SingleSegmentImpl(resourceDescription, segments[0], length, chunkSizePower, false);
+    } else {
+      return new MultiSegmentImpl(resourceDescription, segments, 0, length, chunkSizePower, false);
+    }
+  }
+
+  private MemorySegmentIndexInput(
+      String resourceDescription,
+      MemorySegment[] segments,
+      long length,
+      int chunkSizePower,
+      boolean isClone) {
+    super(resourceDescription);
+    this.segments = segments;
+    this.length = length;
+    this.chunkSizePower = chunkSizePower;
+    this.chunkSizeMask = (1L << chunkSizePower) - 1L;
+    this.isClone = isClone;
+    this.curSegment = segments[0];
+  }
+  
+  void ensureOpen() {
+    if (curSegment == null) {
+      throw alreadyClosed();
+    }
+  }
+
+  RuntimeException wrapAlreadyClosedException(RuntimeException e) {
+    if (e instanceof NullPointerException) {
+      return alreadyClosed();
+    }
+    // TODO: maybe open a JDK issue to have a separate, more
+    // meaningful exception for unmapped segments:
+    if (e.getMessage() != null && e.getMessage().contains("closed")) {
+      return alreadyClosed();
+    }
+    return e;
+  }
+  
+  // return value only for compiler!
```

Comment:
Is that a JDK16 thing? It thinks this is a no-op?

Reply:
no, we need some return value so we can call it and return a value. But as it is throwing an exception anyways, the return value does not matter.

These tricks are often used for exception handlers to delegate "exception handling" to a method where you know it will never return and just bubble up the exception:
- if the exception is known before (by type), you use `throw callSomeMethod(param)`. The pro thing: it's clean as you delegate the instantiation of the exception to the factory, but throw it on your own. The problem here is: only works with a known exception type - and a single one, because the method needs it as return type. A generic `Exception` is a desaster, as it reuires to declare that on throws! 
- an alternative is (and that's used here): put the throw clause also into the method, so the only sense of our method is throwing some exception, it never returns cleanly. Backside: the method needs a return type, so you can do: `return callSomeMethod()` from the caller. Of course you can add a separate return statement after the method call, but that looks more bullshitty.

Here we use the second approch. The most common return type to all callers of this method is `byte`. Compiler is happy, voila!

Reply:
See here: https://github.com/apache/lucene-solr/pull/2176/files#diff-1c8534e50633cd8648f2cb3c886a269a6ebf8a57d817885bd962225d83f4e93fR450

`byte` is fine for a method that can return `long`, `int`, `short`, `byte`,.... or `void`.

Reply:
I see, it saves the (bullshitty) pretense of returning some dummy value when that case will never actually occur - neat